File: TEST_RESOURCE/memory_vulns.cpp
#include <iostream>
#include <cstring>
#include <memory>
#include <vector>

class VulnerableClass {
public:
    int* data;
    
    VulnerableClass(int size) {
        data = new int[size];
    }
    
    ~VulnerableClass() {
        delete[] data;
    }
};

// 1. BUFFER OVERFLOW
void buffer_overflow_cpp() {
    char buffer[10];
    std::string large(100, 'A');
    strcpy(buffer, large.c_str());  // VULN: buffer overflow
}

// 2. USE AFTER FREE (raw pointer)
void use_after_free_raw() {
    int* ptr = new int(42);
    delete ptr;
    *ptr = 100;  // VULN: use after free
}

// 3. USE AFTER FREE (smart pointer misuse)
void use_after_free_smart() {
    std::unique_ptr<int> ptr = std::make_unique<int>(42);
    int* raw = ptr.get();
    ptr.reset();
    *raw = 100;  // VULN: use after free via raw pointer
}

// 4. DOUBLE FREE
void double_free_cpp() {
    int* ptr = new int(42);
    delete ptr;
    delete ptr;  // VULN: double free
}

// 5. MEMORY LEAK (raw pointer)
void memory_leak_raw() {
    int* leak = new int[1000];
    // VULN: never deleted
    leak[0] = 1;
}

// 6. MEMORY LEAK (exception safety)
void memory_leak_exception() {
    int* leak = new int[1000];
    throw std::runtime_error("exception");  // VULN: leak if exception thrown
    delete[] leak;
}

// 7. NULL DEREFERENCE
void null_dereference_cpp() {
    int* ptr = nullptr;
    *ptr = 42;  // VULN: null dereference
}

// 8. DANGLING REFERENCE
int& dangling_reference() {
    int local = 42;
    return local;  // VULN: returning reference to stack
}

// 9. ARRAY OUT OF BOUNDS
void array_out_of_bounds() {
    std::vector<int> vec(5);
    for(int i = 0; i <= 5; i++) {
        vec[i] = i;  // VULN: out of bounds at i=5 (debug mode may crash)
    }
}

// 10. INVALID ITERATOR
void invalid_iterator() {
    std::vector<int> vec = {1,2,3,4,5};
    auto it = vec.begin();
    vec.push_back(6);  // may reallocate
    *it = 10;  // VULN: iterator invalidated
}

// 11. DOUBLE DELETE (shared_ptr cycle not really, but raw delete)
void double_delete_raw() {
    int* ptr = new int(42);
    delete ptr;
    delete ptr;  // VULN: double delete
}

// 12. VIRTUAL DESTRUCTOR MISSING
class Base {
public:
    ~Base() {}  // VULN: non-virtual destructor
};

class Derived : public Base {
public:
    int* buffer;
    Derived() { buffer = new int[100]; }
    ~Derived() { delete[] buffer; }
};

void missing_virtual_dtor() {
    Base* obj = new Derived();
    delete obj;  // VULN: Derived destructor not called, memory leak
}

int main() {
    std::cout << "=== MEMORY VULNERABILITY TEST SUITE (C++) ===" << std::endl;
    
    buffer_overflow_cpp();
    use_after_free_raw();
    use_after_free_smart();
    double_free_cpp();
    memory_leak_raw();
    null_dereference_cpp();
    array_out_of_bounds();
    invalid_iterator();
    double_delete_raw();
    missing_virtual_dtor();
    
    int& ref = dangling_reference();
    std::cout << ref << std::endl;  // VULN: dangling reference
    
    return 0;
}

File: TEST_RESOURCE/memory_vulns.rs
// RUST - Les vulnérabilités mémoire nécessitent 'unsafe'
// Ces exemples sont pour TESTER votre analyseur

// 1. BUFFER OVERFLOW (via unsafe)
fn buffer_overflow() {
    unsafe {
        let mut buffer = vec![0u8; 10];
        let large = vec![65u8; 100];  // 'A' * 100
        std::ptr::copy_nonoverlapping(
            large.as_ptr(),
            buffer.as_mut_ptr(),
            100  // VULN: writing past buffer bounds
        );
    }
}

// 2. USE AFTER FREE (via raw pointer)
fn use_after_free() {
    let ptr = Box::into_raw(Box::new(42));
    unsafe {
        drop(Box::from_raw(ptr));  // free
        println!("{}", *ptr);  // VULN: use after free
    }
}

// 3. DOUBLE FREE
fn double_free() {
    let ptr = Box::into_raw(Box::new(42));
    unsafe {
        drop(Box::from_raw(ptr));
        drop(Box::from_raw(ptr));  // VULN: double free
    }
}

// 4. MEMORY LEAK (forget)
fn memory_leak() {
    let leak = Box::new(42);
    std::mem::forget(leak);  // VULN: memory leak
}

// 5. MEMORY LEAK (cycle with Rc)
use std::rc::Rc;
use std::cell::RefCell;

struct Node {
    next: RefCell<Option<Rc<Node>>>,
}

fn memory_leak_cycle() {
    let first = Rc::new(Node { next: RefCell::new(None) });
    let second = Rc::new(Node { next: RefCell::new(None) });
    *first.next.borrow_mut() = Some(Rc::clone(&second));
    *second.next.borrow_mut() = Some(Rc::clone(&first));
    // VULN: reference cycle - memory leak (needs weak)
}

// 6. NULL DEREFERENCE (via raw pointer)
fn null_dereference() {
    let ptr: *const i32 = std::ptr::null();
    unsafe {
        println!("{}", *ptr);  // VULN: null dereference
    }
}

// 7. DANGLING POINTER
fn dangling_pointer() -> *const i32 {
    let local = 42;
    &local as *const i32  // VULN: returning pointer to stack
}

// 8. OUT OF BOUNDS (safe Rust panics, but unsafe bypasses)
fn out_of_bounds_unsafe() {
    let vec = vec![1,2,3,4,5];
    unsafe {
        println!("{}", vec.get_unchecked(10));  // VULN: out of bounds
    }
}

// 9. UNINITIALIZED MEMORY READ
fn uninitialized_read() {
    unsafe {
        let ptr: *mut i32 = std::alloc::alloc(
            std::alloc::Layout::new::<i32>()
        ) as *mut i32;
        println!("{}", *ptr);  // VULN: reading uninitialized memory
        std::alloc::dealloc(ptr as *mut u8, std::alloc::Layout::new::<i32>());
    }
}

// 10. MISMATCHED ALLOCATION/DEALLOCATION
fn mismatched_alloc() {
    unsafe {
        let ptr = std::alloc::alloc(std::alloc::Layout::new::<i32>());
        drop(Box::from_raw(ptr as *mut i32));  // VULN: mismatched free
    }
}

fn main() {
    println!("=== MEMORY VULNERABILITY TEST SUITE (RUST UNSAFE) ===");
    
    buffer_overflow();
    use_after_free();
    double_free();
    memory_leak();
    memory_leak_cycle();
    null_dereference();
    out_of_bounds_unsafe();
    uninitialized_read();
    
    let dangling = dangling_pointer();
    unsafe { println!("{}", *dangling); }  // VULN: dangling pointer
}